content/src/debug.js

```diff
@@ -201,38 +364,205 @@ function reportSeeeval(method, debugId, length, coordId, elem, args){
   debugRecordsByDebugId[currentDebugId] = record;
 }
 
+function reportEnter(method, debugId, length, coordId, elem, args){
+  stuckComplexity.calls += 1;
+  var recordD = debugRecordsByDebugId[debugId];
+  var recordL = debugRecordsByLineNo[recordD.line];
+  recordD.animated = false;
+  recordL.animated = false;
+}
+
+
 function reportAppear(method, debugId, length, coordId, elem, args){
-  stuckComplexity.moves += 1;
+if (!programChanged) { 
   var recordD = debugRecordsByDebugId[debugId];
-  if (recordD) {
-    if (!recordD.seeeval) {
-      var recordL = debugRecordsByLineNo[recordD.line];
-      recordD.method = method;
-      recordL.method = method;
-      recordD.args = args;
-      recordL.args = args;
+  var recordL = debugRecordsByLineNo[recordD.line];
+  recordD.method = method;
+  recordL.method = method;
+  recordD.args = args;
+  recordL.args = args;
+
+  var currentRecord = debugRecordsByDebugId[currentRecordID];
+  var currentIndex = currentRecord.eventIndex;
+  var currentLocation = traceEvents[currentIndex].location;
+  var currentLine = currentLocation.first_line;
+
+  var prevLine = 0;
+
+  stuckComplexity.moves += 1;
+
+  if (recordD) { 
+    if (!recordD.seeeval){ 
       var index = recordD.eventIndex;
-      var location = traceEvents[index].location.first_line;
+      var line = traceEvents[index].location.first_line;
+      var appear_location = traceEvents[index].location;
+      var tracedIndex = -1; 
+
+      //trace lines that are not animation.
+      while (currentRecordID < debugId) {
+        if (prevIndex != -1) {
+          var prevLocation = traceEvents[prevIndex].location;
+          prevLine = prevLocation.first_line;
+        }
+        else {
+          var prevLocation = null;
+          prevLine = -1;
+        }
+
+        if (tracedIndex!= -1) {
+          untraceLine(tracedIndex);
+          tracedIndex = -1;
+        }
+
+        if(currentLine < prevLine && currentIndex == prevIndex + 1) {
+
+          if (arrows[prevIndex] != null) {
+            arrows[prevIndex]['after'] =  {first: currentLocation, second: prevLocation};
+          }
+          else {
+            arrows[prevIndex] = {before: null, after: {first: currentLocation, second: prevLocation}};
+          }
+          if (arrows[currentIndex] != null) {
+            arrows[currentIndex]['before'] =  {first: currentLocation, second: prevLocation};
+          }
+          else{
+            arrows[currentIndex] = {before: {first: currentLocation, second: prevLocation}, after : null};
+          }
+          view.arrow(view.paneid('left'), arrows, currentIndex, false);
+          debugRecordsByLineNo[prevLine].eventIndex = prevIndex;
```

Comment:
Yup, here it's confusing to me that we write anything into debugRecordsByLineNo.  The map debugRecordsByLineNo should really have just one purpose, which is to support by-line hovering.  The main collection is debugRecordsByDebugId, and we should be manipulating things in there.  (debugRecordsByLineNo is just a view into that collection.)